services/server/src/server/controllers/repository/repository.handlers.ts

```diff
@@ -128,9 +136,71 @@ export async function checkAllByChainAndAddressEndpoint(
             });
           }
 
-          map
-            .get(address)
-            .chainIds.push({ chainId, status: found[0].runtimeMatch });
+          // Proxy detection and resolution
+          type Implementation = { address: string; name?: string };
+          let proxyStatus: {
+            isProxy?: boolean;
+            proxyType?: ProxyType | null;
+            implementations?: Implementation[];
+            proxyResolutionError?: string;
+          } = {};
+          if (
+            req.query.resolveProxies === "true" &&
+            found[0].onchainRuntimeBytecode
+          ) {
+            try {
+              const sourcifyChain = chainRepository.supportedChainMap[chainId];
+              const proxyDetectionResult = await detectAndResolveProxy(
+                found[0].onchainRuntimeBytecode,
+                address,
+                sourcifyChain,
+              );
+
+              const implementations = await Promise.all(
+                proxyDetectionResult.implementations.map(
+                  async (implementationAddress) => {
+                    implementationAddress = getAddress(implementationAddress);
+                    const implementation: Implementation = {
+                      address: implementationAddress,
+                    };
+
+                    const implementationFound: Match[] =
+                      await services.storage.performServiceOperation(
+                        "checkAllByChainAndAddress",
+                        [implementationAddress, chainId],
+                      );
+                    if (
+                      implementationFound.length > 0 &&
+                      implementationFound[0].contractName
+                    ) {
+                      implementation.name = implementationFound[0].contractName;
+                    }
```

Comment:
From what I understand we only return `contractName` only if the implementation is verified on Sourcify. Otherwise we just return the address. 

If so we should note this down in the documentation or maybe have a separate field for `isVerified`